stdlib/public/core/CharacterProperties.swift

```diff
@@ -48,11 +56,65 @@ extension Character {
   ///     // lf.asciiValue == 10
   ///     let crlf = "\r\n" as Character
   ///     // crlf.asciiValue == 10
-  @inlinable
+  @_alwaysEmitIntoClient @_silgen_name("inlineASCIIValue")
   public var asciiValue: UInt8? {
-    if _slowPath(self == "\r\n") { return 0x000A /* LINE FEED (LF) */ }
-    if _slowPath(!_isSingleScalar || _firstScalar.value >= 0x80) { return nil }
-    return UInt8(_firstScalar.value)
+    if !_str._guts.isSmall { return outlineASCIIValue }
+    // This character is backed by a small string. We can check if it is a
+    // single ASCII character (and return its value) just by looking at the
+    // first word. This works because Character is made up of a single
+    // grapheme cluster, the NUL character does not combine, and in a
+    // SmallString all unused bytes must be zero, so we don't need to look
+    // at the count.
+    let firstWord = _str._guts.asSmall._storage.0.littleEndian
+    return firstWord < 128 ? UInt8(firstWord) : firstWord == 0x0a0d ? 0xa : nil
+  }
+  
+  // Mangle this as "isASCII" for ABI compatability with older libraries where
+  // the generic isASCII was inlinable public (so calls will exist, even though
+  // it was usually inlined).
+  @usableFromInline @_silgen_name("$sSJ7isASCIISbvg") @inline(never)
+  internal var outlineIsASCII: Bool {
+    // On binary-stable platforms, older binaries might reference this symbol
+    // even when the backing string is small, so we still need to handle that
+    // here.
+    // TODO: this could be eliminated on non-binary-stable platforms
+    if _str._guts.isSmall { return isASCII }
+    // A non-small Character can only be tail-allocated native, so we can
+    // get the count and pointer without the usual dispatch.
+    let count = _str._guts._object.largeCount
+    let pointer = _str._guts._object.nativeUTF8Start
+    if count == 1 {
+      return pointer[0] < 128
+    }
+    if count == 2 {
+      return UInt16(pointer[0]) | UInt16(pointer[1]) << 8 == 0x0a0d
+    }
```

Comment:
Were you able to construct an example where the large-string case ever returned true? That is, ASCII can be up to 2 UTF-8 code units and we might only ever form small strings in that case. This depends on whether we use always  construct new storage for each `Character` or if there are some code paths where a 1-count `String`'s guts would be reused.

Reply:
I was not, but neither did I do the analysis to convince myself that it cannot happen.